brion/nodeGroup.cpp

```diff
@@ -0,0 +1,185 @@
+/* Copyright (c) 2013-2018, EPFL/Blue Brain Project
+ *                          Jonas Karlsson <jonas.karlsson@epfl.ch>
+ *
+ * This file is part of Brion <https://github.com/BlueBrain/Brion>
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "nodeGroup.h"
+
+#include <HighFive/include/highfive/H5File.hpp>
+#include <HighFive/include/highfive/H5Group.hpp>
+
+namespace brion
+{
+struct NodeGroup::Impl
+{
+    HighFive::Group group;
+};
+
+NodeGroup::NodeGroup(const HighFive::Group& group)
+    : impl(new NodeGroup::Impl())
+{
+    impl->group = group;
+}
+NodeGroup::~NodeGroup() = default;
+
+Strings NodeGroup::getAttributeNames()
+{
+    Strings attributeNames;
+    const size_t num_objects = impl->group.getNumberObjects();
+
+    for (size_t i = 0; i < num_objects; i++)
+        attributeNames.push_back(impl->group.getObjectName(i));
+
+    return attributeNames;
+}
+
+Strings NodeGroup::getDynamicParameterNames()
+{
+    Strings parameterNames;
+
+    const HighFive::Group dynamics_params_group =
+        impl->group.getGroup("dynamics_params");
+
+    const size_t num_objects = dynamics_params_group.getNumberObjects();
+
+    for (size_t i = 0; i < num_objects; i++)
+        parameterNames.push_back(dynamics_params_group.getObjectName(i));
+
+    return parameterNames;
+}
+
+size_t NodeGroup::getNumberOfNodes() const
+{
+    return impl->group.getNumberObjects();
+}
+
+template <typename T>
+T getAttributeHelper(const HighFive::DataSet set, bool use_slice = false,
+                     size_t start = 0, size_t end = 0)
+{
+    T list;
+    set.read(list);
```

Comment:
This is not efficient for the slice case, you're reading everything. You have to use the function HighFive::SliceTraits::select, which is inherited by HighFive::DataSet